### ISSUE TITLE
Use decentralized discovery (DHT)

### ISSUE THREAD

**Issue by calmh** [syncthing/syncthing#3388]
It's probably time to investigate decentralizing discovery. Doing so would reduce our infrastructure burden and reduce centralized knowledge of the network. In a DHT we want to be able to look up the contact details for a given device ID. We want to publish our own contact details into the DHT. We don't necessarily want to make it simple for a passive observer to map the entire network, so it would be nice if the contact details were encrypted. I suggest that we could use a key-value database laid out like:

Key: SHA256(device ID)

Value: The contact details, in protobuf format, encrypted using the device ID as a shared secret.

Given that I know the device ID I'm looking for, I can then hash it and look it up, and decrypt the answer. For other nodes in the network, the data is mostly gibberish as you need to know the device ID it represents to be able to interpret it.

As far as an actual implementation goes I've surveyed the Go ecosystem to try to find a suitable DHT package. So far I've not really succeeded. Here are the ones I've looked at and some notes on them; it would be lovely if someone did find or develop a good package for this.

- https://github.com/nictuku/dht seems to be the best candidate, but a partial implementation - only supports looking up nodes for an "info hash", not storing or retrieving actual values? maybe could be built upon
- https://github.com/distributed/malus toy project, last commit 6 years ago
- https://github.com/cfromknecht/kademlia partial implementation, uses TCP?
- https://github.com/worldveil/peerchat student project, abandoned?, no license
- https://github.com/zhujun1980/dhtrobot looks sort of promising, need to get past the chinese docs
- https://github.com/gr0gmint/godht incomplete, abandoned

**Comment by AudriusButkevicius** [comment]
Ipfs people are using kademila afaik. 

Anyways, you still need infrastructure to bootstrap, and how do we perform authentication against other peers being non-malicious, prevent malicious peers from overwriting keys, or even bootstrap server from being malicious?

**Comment by calmh** [comment]
https://godoc.org/github.com/ipfs/go-ipfs/routing/dht looks quite good, yes, thanks!

Certainly we'll need a couple of bootstrap nodes. But I don't think they need to be as beefy as our current discovery servers, and more importantly they shouldn't need to scale much in the future.

As for the maliciousness I'm not sure. Decentralizing things makes that difficult. The worst case scenario is then being tricked to connect to the wrong place when looking for a certain device ID, blocking discovery of a given device...

**Comment by raichu** [comment]
Also relevant:

http://blog.notdot.net/2009/11/Implementing-a-DHT-in-Go-part-1
http://blog.notdot.net/2009/11/Implementing-a-DHT-in-Go-part-2

A note about Kademlia's DoS resistance:
https://github.com/ethereum/wiki/wiki/Kademlia-Peer-Selection#backwards-compatibility-mode-and-dos-safety

A security-enchanced version of Kademlia resistant against various types of DoS attacks (described in section 3.2): http://www.tm.uka.de/doc/SKademlia_2007.pdf

**Comment by cdhowie** [comment]
> The worst case scenario is then being tricked to connect to the wrong place when looking for a certain device ID, blocking discovery of a given device...

I don't think that's a feasible attack, since the value would be encrypted by the device ID.  Such an attack would have to be targeted against a specific, known device ID -- and for that to even work, your DHT node needs to be the one responsible for SHA256(device ID).  And even if you successfully pull that off, the traffic is end-to-end encrypted.  All you'd really be able to see (assuming you proxy the data) are the IPs on both sides and the amount of data being transferred.  You can kinda do this already by running a public relay server, right?

The "store garbage" attack is much more worrying to me.  To my knowledge, the way other DHT networks (torrents for example) combat this problem is by having the key be a hash of the value -- not something we can realistically do here.

**Comment by cdhowie** [comment]
> The "store garbage" attack is much more worrying to me.

A realization I just had is that any kind of verification mechanism put in place to avoid a garbage attack would have to incorporate a timestamp, or it would be susceptible to a replay attack, where a previous (valid) value is stored under the same key.  For mobile clients that are constantly changing IPs, this could prevent any nodes from connecting to them.

**Comment by calmh** [comment]
I think this will necessarily be some sort of security trade off. We'll gain some security (no need for a trusted middle man who has all the data, harder to DoS, etc) and lose some (harder to prevent false information). But as long as the feasible attack is mostly to prevent lookup against a specific device ID I think it may be worth it. It's just one discovery method among many after all.

**Comment by AudriusButkevicius** [comment]
Or just DDoS the bootstrapping node...

**Comment by calmh** [comment]
That could still be fine though, if you could still reach some other device to bootstrap, say via local discovery, or if you're already connected to the network. But sure, I'd still agree we need the three geographically diverse bootstrap nodes and it's certainly not bulletproof. But honestly the bigger threat at the moment is probably me screwing up the discovery servers by fatfingering something. :)

**Comment by aviau** [comment]
> That could still be fine though, if you could still reach some other device to bootstrap, say via local discovery, or if you're already connected to the network

One other method around DDoS-ing the bootstrap node is that clients should save all nodes that they know to the disk. Whenever they restart they can try to bootstrap with all of these nodes. The bootstrap node is only critical for the fist time the client is started.

**Comment by Ferroin** [comment]
That becomes problematic when dealing with clients with dynamic addresses, you'd need a system with a static address in the network for such a scheme to work reliably, because there's the possibility that all the other systems change address.

**Comment by canton7** [comment]
I think @aviau meant all other DHT nodes which that device had discovered, and not all configured Syncthing devices.

There will probably be enough DHT nodes which either have static IPs, or whose IP hasn't changed (most dynamic IPs rarely change in practice) for a node to reconnect itself to the DHT network.

**Comment by Ferroin** [comment]
To retain feature parity though, there would need to be an option to run independently of the default bootstrap servers and  DHT network.  Otherwise, this is functionally a regression for anyone using a custom global discovery server.  This becomes even more important if all nodes are expected to participate, especially considering people like me who have to pay for bandwidth and data transfer for our nodes which do have static addresses.  Such a situation of paying for bandwidth is a lot more common than you think, an will be the case for almost anyone using cloud hosting or running over a mobile broadband connection, and in fact is much better served by a centralized discovery mechanism like what Syncthing has now.

**Comment by aviau** [comment]
> There will probably be enough DHT nodes which either have static IPs

Exactly. We use this method for Ring.

**Comment by masgo** [comment]
Maybe a "traditional" DHT is not a good solution. You would have to extend it to have something like aging (for IDs that went away and which do not come back). This would need 

Also, most DHTs do not behave well with high churn rates. With today's mobile devices you will have a lot of churn because of unstable mobile connections, unstable WiFi, power-save features of the devices or non-24/7 dial-up connections.

Why don't you start with the low hanging fruits? Peer Exchange! (not necessarily the PEX implementation used by BitTorrent - it also has some drawbacks). The idea would be: enable all devices as discovery/relay server. After initial discovery (via your servers) the devices would know 1 or multiple other devices. The next time they connect, they would try to reach the same address. If one of the old addresses is still working (e.g., because the device has a fixed IP, or because it did not go offline in the meantime) they would connect to it, and this device would introduce them to the other devices.

Thinking of the UI:
You could introduce a simple switch like "this node uses a metered connection, use only as much bandwidth as necessary". This would disable the node being a relay, it would exclude it from the DHT, and maybe even disable it being a discovery server.

**Comment by calmh** [comment]
I suspect that PEX might not be as useful for us as it is for BitTorrent. In the BT case you have hundreds or thousands of devices that are all interested in the same resource - a given torrent. They already need to track other peers interested in the same thing (for downloads) so it's no extra load to provide this information to other devices as well.

In our case, most devices are not interested in the devices you want to talk to so they won't have the information available.

**Comment by masgo** [comment]
I agree, but it would work great for every (individual) group of devices which use the same shared folder. IF there is either (a) at least one device with a fixed IP, or (b) at least one device keeps its IP between sessions of the other devices. 

If I look at the statistics about syncthing, 18% use the "static addresses" feature and the "Number of Devices in Cluster" is mostly 3 or more. 

So, using any device of the group as discovery server and simply contacting all previously known devices would solve the discovery problem for ~18% of the users (or even more). It would also work great in large LAN scenarios where IPs tend to stay the same, but broadcast (and therefore Local Discovery) might be filtered.

The next scenario which I would expect to be quite common is a home setup where one device runs most of the time. The external IP will change from time to time (once/day). If a mobile device would simply try to connect to last know IP it has a very high probability to succeed. 

My argument is: it would be cheap to simply try all previously known IPs before contacting a discovery service. If we also save the previously measured RTT, one could wait (for example) 2xRTT before contacting the global discovery.

My approach will certainly not fully replace global discovery, but it might reduce the load greatly.

**Comment by Ferroin** [comment]
Adding on to @masgo's idea: Given a situation with three nodes A, B, and C, it's likely that all three want to share data directly.  If B has a fixed address, then it makes sense for it to act as a discovery server among it's peers (for example, A could use B to get info about how to talk to C).  If this gets extended to locally discovered peers (ie, if B is on the same network as A, and is already talking to C, then A could get info from B about how to talk to C), it becomes a context limited version of PEX.

**Comment by Ferroin** [comment]
@calmh Also, your argument against using PEX could be applied to a DHT setup as well:  All my devices care about is how to talk to the rest of my devices, why should I use my bandwidth and processing power to help your devices find their peers?

**Comment by AudriusButkevicius** [comment]
And how does the existing discovery system not solve the current problem?
The only beneift of the whole discussion I see so far, is the ability to connect when discovery servers are down, one of the peers has a static address and you happen not to be in LAN, essentially by storing last N addresses for a peer, amd sharing addresses of peers you are aware the other side is interested in after connect.

**Comment by masgo** [comment]
> And how does the existing discovery system not solve the current problem?

The existing system solves this with the global discovery. This is about

> reduce our infrastructure burden

**Comment by AudriusButkevicius** [comment]
I was referring to DHT needing extra bandwidth comment above, not in general, yet that was used as a counter argument.

**Comment by calmh** [comment]
> My argument is: it would be cheap to simply try all previously known IPs before contacting a discovery service.

Absolutely agree. We should do this.

> All my devices care about is how to talk to the rest of my devices, why should I use my bandwidth and processing power to help your devices find their peers?

Because then you get to use the distributed discovery mechanism to find your other devices :) In my vision of the DHT each device only keeps a small part of the whole map, so the burden shouldn't be undue.

**Comment by Ferroin** [comment]
@calmh the problem I personally have with that is that there's a lot of people who pay per unit transferred or time used for their network access.  It's also not unusual to have transfer caps.  Having it as an option is not a bad thing, but making it mandatory is, people will suddenly see Syncthing eating up more of their network bandwidth than before, and potentially costing them more money to use than before.

If we were to implement retrying previously seen addresses prior to discovery, and then have peers share discovery info about their other peers, then the discovery servers would only be needed for bootstrap of a group or for cases when all addresses changed (and possibly for finding relays).

**Comment by calmh** [comment]
Ah, I think I see what you're getting at now. Having devices within cluster share address information about other devices in the same cluster. We should do that of course. I was thinking you meant some sort of exchange with devices that had nothing to do with me. Sorry for being slow.

**Comment by subscribernamegoeshere** [comment]
Hi there I probably know way too little about these exact technical details, but what I have been wondering for quite a number of years now, that when during all these years people came up with new incarnations and implementations of their application and networks and distributed stuff, why not use already existing and well estblished dht out there in kind of a parasitic or freeloader freeriding or better say symbiontic ways. For example the bittorrent guys original mainline dht implementation probably the largest biggest and badassiest dht with tens of millions of participating nodes I guess. Azureus/Vuze, vanilla bittorrent client or utorrent and all those use that network for bootstrapping and finding fellow nodes. If I understand this at all, why not implement or use an implementation that takes part in this dht network and then maybe add to it for detecting which nodes once we use that, are offering the requested services, meaning are for example not actual bt nodes but syncthing nodes and so on. Maybe this is all extensible and quite modular for smybiontic and friendly use?

See: <http://engineering.bittorrent.com/2013/12/19/dht-bootstrap-update/> http://engineering.bittorrent.com/2013/12/19/dht-bootstrap-update/

Everybody and their brother seem to come up with the need to maintain a dht of their own every few weeks or months, so why not cooperate and all be in one giant dht that can not be taken down and that will or is not only known for copyright trouble or torrenting but for generic use to find nodes in a giant cloud of nodes and their services to be distinguished further down the road after initial lookup inside dht.

Any insightful commands? Does bittorrent.com guys actually allow for their dht to be used? For non-torrenting needs to join their dht maybe as well? Anyone officially talking to them? I remember some long while ago when these similar questions and problems with for example opensource application/project retroshare came up and they actually used a vanilla style bittorrent dht then but created it disjointly from the bittorrent.com guys instead of using and bootstrapping from the existing one already.

Thanks.

**Comment by Ferroin** [comment]
@subscribernamegoeshere I assume you mean aside from the fact that doing so would be a breach of the TOS of all such services, and therefore functionally illegal in almost all jurisdictions worldwide?

In most cases, BitTorrent clients don't even talk to anything owned by the company that owns the copyrights.  Most people who serve torrents run their own trackers (which are the bit that handles bootstrapping the DHT network), so there's not anything going back to bittorrent.com most of the time.  

Additionally, DHT, like almost any other distributed system, is application specific.  DHT as used in BitTorrent wouldn't work for us because we need a different key<->value mapping.

And then there's the fact that DHT is a high-level design.  The concept is used in many things, but not all of them can understand each other.  DHT as used by BitTorrent is not the same as DHT as used by ipfs for example, and neither would be the same as how Syncthing would use it (BitTorrent and IPFS use different implementations for content lookup, Syncthing would be using it for device discovery).

**Comment by subscribernamegoeshere** [comment]
I mean the lowest levels of dht finding other nodes for bootstrapping. I am not talking about torrenting. Read the engineering blog of theirs and look at retroshare, which actually does use the bittorrent dht "cloud" of nodes out there and asks that dht for node addresses to find other nodes and only then tries to talk the retroshare language or retroshare dialect of stuff to the found nodes. I dont exactly know on a tech level how it differentiates the found nodes via the dht if those nodes would be torrent clients or retroshare clients. Thats what I am talking about. I dont know about legality or TOS or whatever, but it would be the task for the developers here to talk to the bittorrent engineers or founders or responsible people out there if their dht can be used in a generic way just to find fellow dht participants. That is my main idea. There are at least two kind of completely disjunct nodes to be found in the bittorrent dht as of today. The torrenting nodes and the retroshare nodes. This a fact. Look into the reotroshare souces how they do it and what exactly they do. DHT doesnt mean torrenting or copyright infringement or these things as per se.

**Comment by subscribernamegoeshere** [comment]
as you say exactly, use the probably largest existing and constantly available dht cloud of nodes out there, the bittorrent.com dht, for simply finding nodes. same way as retroshare does. that is what i mean. not to find content of files or anything else, but to merely find fellow nodes, fellow syncthing participants. retroshare does find other retroshare participants quickly this way, so there must be some built in or easily extensible way to filter out or to find the non-torrenting nodes on the bittorrent dht in some way, to distringuish the various type of nodes offering various types of services (e.g. torrenting, non-torrenting, retroshare, syncthing, $yourservice)

**Comment by AudriusButkevicius** [comment]
As we said, DHT doesn't really work until we iron out tye details, as there is no authentication. Someone can prevent you from connecting by constantly overwriting the key that stores your location with garbage. Or can overwrite all the syncthing keys in the DHT and cause a DDoS on someone.

About reusing someone elses DHT, I am also against it as it it puts control in someone elses hands.
What if bittorrent people decide that syncthing is causing too much competition to btsync and fuck us over by refusing syncthing connections on the bootstrap nodes?

**Comment by subscribernamegoeshere** [comment]
Uhm how does bittorrent bootstrapping then works if any malicious node could simply pollute their dht. Something must be wrong here with reasoning. I am not an expert but their dht and source code is documented so how come the mpaa or someone else has never took down their dht if it was possible at all? Please explain dht to me then maybe I am not understand it or we all have weird misconceptions of it. You can not just write to any arbitrary key on a dht, I think that was an essential basic write and read option back in the days of opendht on planetlabs, they offered two python or something scripts, the one wrote to the dht and the other simply read from it. You protect the write operation of a key by adding a salt or a password, and only via that you can (re)write that key/value, but anyone can read a key/value. Eventually a key/value timesout or gets deprecated and garbagecollected e.g. long time no see host/node/ip/port combination and so on. Am I completely misunderstanding the dht concept and how you would put up a pretty basic and stupid protection of added values? Anyone?

**Comment by AudriusButkevicius** [comment]
So I don't claim that I understand it works, I claim that I haven't seen anyone answering my question about key overwrites, hence why I raised the point again before we dvelve into which network to use.

As far as I understood, and I might be wrong, was that keys were append only and that was the only protection.

**Comment by masgo** [comment]
@subscribernamegoeshere
> Everybody and their brother seem to come up with the need to maintain a dht of their own every few weeks or months, so why not cooperate and all be in one giant dht

By now, I have done some years of research on Peer-to-Peer and distributed systems. I came across many DHT and other systems with similar properties (like distributed rendevous systems), I have also implemented some DHTs. The short answer to your question is: because systems have different requirements and each DHT has a lot of (tuning) parameters. 

Some details where systems differ:
- How is the key determined (fron IP+Port, random, or by some other means)
- How much data will be stored along each key: a few KBs (for example routing information) or some MBs or even a lot more (like GBs, TBs or even PBs)
- How stable/reliable are the participants? Will they be only for long periods of time (usually the case with cloud instances) or will they come and go (like in torrent clients) or will they switch between online and offline often (like mobile clients). This is called "**churn**". Some well known systems (like _Chord_) are known to deal badly with high churn rates, while other will work great even if 50% of the participants crash at once (like [BubbleStorm](https://github.com/DVS-P2P/bubblestorm), developed by my former colleagues) 
- How fast should the lookup be? Most DHTs provide `log(number of participants) ` lookup performance. But there is a difference between log_2 and log_10. One might also optimize in such a way that each step has a lower RTT (instead of the average RTT between all participants). Should each lookup have (roughly) the same lookup time, or do we optimize for faster lookups of popular items?
- What kind of lookup operations do we support? Only key-value (like many "traditional" DHTs), range queries on the key, multi-key lookup, keyword search on the data or even arbitrary queries (SQL-like queries or RegExp) on the data?
- Each system will cause an overhead in bandwidth. How high can this overhead be? Should the bandwidth consumption be the same for all nodes or will nodes with higher bandwidth also make use of it?
- What replication model/properties do I want? Will it be persistent? Should it be short-term or long-term storage? Should only the original publisher be responsible for it and will it be deleted if he leaves, or will it be independent of the publisher?
- Who distributes the query? Will the querying node have to contact the second hop or will the first hop contact the second hop?
- And many more
- And last but not least: **security**. This is an orthogonal topic on its own. Security against DoS attacks on the whole system or on single nodes. Security against manipulation of data, etc etc.


# Coming back to Syncthing
Let's start by gathering (and agreeing) on the requirements. 
- What do we want to store?
    - As far as I can tell we could store (Device-ID, addresses), and/or (Folder-ID, Device-ID, and maybe addresses), or (Cluster-ID, addresses of all devices) tuples. 
- We probably want time-stamps with each entry
- What security properties do we want for the data? 
    - Which information should be kept secret? And from whom?
    - Which should be protected (changes only possible by original author)?
    - What about protection against data-loss? (This is usually a trade-off between re-insert frequency and replication)
- What protection do we want to offer to the participants?
   - Limit the amount of stored data?
   - Limit the used bandwidth
- Do we tolerate false positives and/or false negatives? 
- What are our key performance metrics?
    - Lookup time?
    - Bandwidth usage?
    - Lookup success rate? false positives, false negatives, ....

**Comment by alugarius** [comment]
Working together with ipfs could not only solve our DHT problem, we could enable the most efficient file shareing ever

**Comment by aviau** [comment]
IPFS uses a DHT, but that dosen't "solve our DHT problem" as there will be coding to do.

Also, I don't think IPFS support sharing private files?

I really don't see how the two could work together.

**Comment by alugarius** [comment]
Every dht solution needs coding
ipfs is actually used to share files privately (its kinda build for that o_O)

Sharing ideas and talk to each other is a kind of working together,
We don't have to use ipfs technology but the team could help us to find something. ...

I don't know where the problem is to try it at least.... we are one community at all!

**Comment by masgo** [comment]
I agree, talking to them can do no harm, especially since the project seems to be under active development.

IPFS itself is (more or less) useless for syncthing since it aims at quite different things. But what might be interesting is the P2P library they use: **[libp2p](https://github.com/ipfs/specs/tree/master/libp2p)**

The library combines many known p2p algorithms and there is a [GO implementation](https://github.com/libp2p/go-libp2p). So talking to them might be a good idea. 


What is still needed is answering the questions regarding our requirements i mentioned above. But with a little luck the requirements can be fulfilled by libp2p.

**Comment by alugarius** [comment]
Thanks for agreeing @masgo  ^^

**Comment by mengzhuo** [comment]
These lib looks much better than libp2p
https://github.com/anacrolix/torrent

**Comment by masgo** [comment]
@mengzhuo you are missing the point. This is not (yet) about which library is best, but about **how** to add distributed discovery to syncthing.

**Comment by Avamander** [comment]
https://www.bountysource.com/issues/35750022-use-decentralized-discovery-dht

**Comment by seanthewebber** [comment]
Sorry for being late to the party, folks. Open Whisper Systems faced this problem with the [contact discovery](https://signal.org/blog/private-contact-discovery/) element of their [Signal messaging app](https://signal.org/). They are using Intel [Software Guard Extensions (SGX)](https://software.intel.com/en-us/articles/innovative-technology-for-cpu-based-attestation-and-sealing) to process contact sharing on secure enclaves. Perhaps this concept is an alternative way of approaching DHT on Syncthing?

**Comment by Ferroin** [comment]
Probably not, as I'm pretty sure you can't easily do stuff with SGX from Go (and I reasonably sure to an even greater degree that you can't do SME (AMD's equivalent to SGX) stuff from GO).  In fact, I can't see how they can require that to begin with in Signal, as SGX is only in the current (and possibly previous, I don't recall for certain) generation of Intel CPU's, and AMD's SME is also only in the last few generations, and more importantly, they have to run on ARM chips which don't have such things.

**Comment by masgo** [comment]
@seanthewebber @Ferroin Appart from relying strictly on certain hardware features, what Signal solves with SGX is a completley different problem. 

They faced the problem: Signal uses a **centralized** management server (or multiple servers) which is used by the users to find other Signal users. In order to do so, you have to query the server like: "is the user 123456 a signal user?". By doing so, you expose the users phone nr (or username, email, etc) to the server as well as the fact that you know this user. A malicious server could use this to collect user-data and build a social graph. This is a big **privacy** issue.

This issue is about **reducing the load on the central infrastructure**. The solution by Signal does not reduce the load on the central server. To the contrary, the load is increased in order to achieve their desired privacy goal.

**Comment by PanderMusubi** [comment]
You can add this line at the top of the issue description, it will update itself

    ![badge](https://api.bountysource.com/badge/issue?issue_id=35750022)

and it looks like

![badge](https://api.bountysource.com/badge/issue?issue_id=35750022)

**Comment by anacrolix** [comment]
You might try [anacrolix/dht](https://github.com/anacrolix/dht) and [BEP 44](https://www.bittorrent.org/beps/bep_0044.html). @mengzhuo kindly mentioned [anacrolix/torrent](https://github.com/anacrolix/torrent) above, the DHT component has since been broken out into a standalone module.

**Comment by calmh** [comment]
Does that package support storing / getting such arbitrary data in the DHT?

**Comment by subscribernamegoeshere** [comment]
its listed in one of their help-needed tagged issues:
https://github.com/anacrolix/dht/issues/26

**Comment by EternityForest** [comment]
I think PEX/Peer Exchange might be more useful here than everyone is suggesting.

Typical home connections don't change IP addresses very often, and sometimes basically never do, for months at a time.

In addition, there is Yggdrasil and CJDNS, which I think are important enough that it's worth adding specific code to be aware of them.

The "Low hanging fruit" approach, to me, would simply be to have nodes try to keep a consistent UPnP external port, and have nodes include their WAN IP/Port in the local discovery packets.

Store that IP, mark it as lower priority than LAN, but when you go off the network, try it, and just remove it if it fails.   95% of the time this will work, the only time it will fail is if the user has not been on the same LAN since the last IP change.

You still need to publish your IP to the servers, but you don't need to look it up except in that 5% of cases, so the load should be halved, since I'm guessing phone-to-PC covers most real use cases.  And, you only very rarely need to publish your IP, since it doesn't change much, and changes can be detected locally via UPnP.

And, even publishing can be avoided.  Before publishing your IP to discovery servers, first attempt to contact all peers.   

If there are no failures, then there is no need to publish your IP,  and any always on nodes that only connect to other always-on nodes will basically never generate traffic.

PEX would let you go slightly further in some scenarios. Since people usually only have a few dozen peers at most, largely overlapping, we can store out IP on *all* of them for a few kb of RAM.

PEX doesn't need to be limited to specific folders, we can just store our info on all of our peers.  DHTs need a lot of round trip traffic, so searching even a dozen peers may well be more efficient.

Peers can also store data indefinitely until rebooted, or even on disk forever, there's no need to refresh or resend data except on initial connection, and you don't even need to explicitly send it, you can use "last seen" info.

And with Yggdrasil or CJDNS, your mesh IP *never* changes, so informing LAN peers of this, and storing it,  is completely sufficient for reliable connection, now that Android supports Yggdrasil, and you have the advantage of being able to operate in a mesh scenario.

I suspect that a combination of these approaches can almost totally eliminate the need for lookups, and bring the system to a fairly high reliability, even if the servers are inaccessible.

Doing the "Low hanging fruit" first, would greatly reduce the reliance on any future DHT, which could get spammed with any amount of garbage.

**Comment by Avamander** [comment]
@EternityForest That suggestion is a topic for some other issue, not this one.

**Comment by masgo** [comment]
@Avamander I think @EternityForest has a point there. There are scenarios where a (modified) PeX would work. Consinder the following:

**Scenario A**
one device is always on --> could keep the same outside port open forever via UPnP. The IP might change but not that often (e.g. daily or monthly). 
In this scenario, if all devices simply keep the last known contact for all their other devices, they could instantly reconnect to the always-on device without using syncthing servers. I.e.: try to connect to all devices via their last known address, most will fail but the always-on device will connect. If the connection fails, remove it from the last known list.
Now, let's add the most simple PeX, where, upon connection, the devices exchange their known other devices. --> The always-on device will basically introduce all other devices. 

**Scenario B**
Among the devices, there is one or more static devices, e.g., PCs. They might not be always on, but they stay in the same (home) network for which the public address rarely changes. Now a device could keep the last (or last n, where n is a low number) known addresses for each device and try them first. (same approach as above). If we become more generous with the removal of addresses (e.g. only remove them after x hours, or y unsuccessful connections), we would still have a pretty good chance to connect to the static devices. Sometimes they will be offline, but sometimes it will work. 

One probably needs to introduce some overall limit on trying peers, in order to avoid being marked as port-scanning by security devices. This could also be combined with a time limit, like: try the peer approach for 5 seconds, then query the syncthing servers while continuing the peer approach. The server response will be put at the front of the "try it" queue.

**Comment by Avamander** [comment]
@masgo, you make two large assumptions a) that the person has a way to set port forwards, b) that at least one device in already discovered. If you lack either, you need to use the centralized discovery methods. Also, again, DHT is what this issue talks about, any talk about PeX and LPD should be in a separate issue.

**Comment by SoniEx2** [comment]
i2p is (effectively) a decentralized STUN/TURN server. i2pd is specifically designed for embedding, too. In addition to being DHT-based, it'd also make syncthing work behind CGNAT. Probably a better option than just a DHT?

**Comment by EternityForest** [comment]
What's the performance drop with i2p?  It would be nice to have the option
of decentralization without the performance hit that anonymity gives.

It seems like a good replacement for relay servers though!

**Comment by SoniEx2** [comment]
Nah, set it to 0 hops. Or rather: keep an endpoint at 1 hop, then exchange the 0-hop endpoint over said 1-hop connection. Attempt connection to the 0-hop endpoint and if that fails just use the 1-hop endpoint.

0 hops = no TURN, direct p2p. 1 hop = TURN. This basically uses i2p for both STUN and TURN and if you happen to get a successful 0-hop connection you don't get any performance drop (other than the crypto but pfft) *and* you don't put any extra load on the i2p network(!). It doesn't provide anonymity, but hey the ISP already knows you use dropbox or syncthing so does it really make a difference?

**Comment by masgo** [comment]
Why not make it configurable for the user with some decent default values for the novice. 
Something like: try syncthing servers first -> if not reachable or overloaded ->use i2p with default settings. 
Then one could simply limit the load, and therefore the cost, on the syncthing servers. 

@SoniEx2 suggestion seems plausible for the novice user. Syncthing has never advertised to be anonymous. 

The crypto overhead might not be negligible for large data transfers and low power/low performance devices (Smartphone, NAS, etc). Especially if one would end in a scenario where data is encrypted multiple times. 

>  but hey the ISP already knows you use dropbox or syncthing so does it really make a difference?

If anonymity is desired: yes it does make a difference. It could help to identify devices which know each other. But as stated above: Syncthing never advertised anonymity as a feature.

**Comment by AudriusButkevicius** [comment]
I don't think this is the right place for discussion about new transports. This issue is specifically about DHT. If you want to have this discussion, please use the forum.

**Comment by Slender1808** [comment]
Having a p2p discovery system improves scalability and also provides backup system for current discovery system as well as allowing users to participate in the project.

My suggestion does not depend on modifications to the project for its functioning.

The way I thought would be by creating a folder and sharing it on syncthing, and somehow getting the id and ip connected in that folder and saving it within itself.

The way to store this data can be through sqlite or folder structure.

### Sqlite structure example:

| ip | id |
|--|--|
| 8.8.8.8 | EQ652TN-45GGV3I-TA6UEPW-VLZBJPD-VBMQXXN-3JTXJ36-4YHLFF7-VTTRQAR |
| 145.46.78.45 | TQ652TN-45GGV3I-TA7UEPW-VLZBJPD-VBMQXXN-3JTXJ36-4YHLFF7-VTTRQAR |

### Folder structure example:
IP 8.8.8.8

/8/8/8.csv

| Class | Initial bits | Start | End | Subnet mask | CIDR notation |
|--|--|--|--|--|--|
| A | 0 | 1.0.0.1 | 126.255.255.254 | 255.0.0.0 | /8 |
| B | 10 | 128.0.0.1 | 191.255.255.254 | 255.255.0.0 | /16 |
| C | 110 | 192.0.0.1 | 223.255.255.254 | 255.255.255.0 | /24 |

using the ip mask as a way to separate into folders and where the last part of the ip mask would be csv file where it would have ip with id

| ip | id |
|--|--|
| 8 | EQ652TN-45GGV3I-TA6UEPW-VLZBJPD-VBMQXXN-3JTXJ36-4YHLFF7-VTTRQAR |
| 45 | TQ652TN-45GGV3I-TA7UEPW-VLZBJPD-VBMQXXN-3JTXJ36-4YHLFF7-VTTRQAR |

**Comment by AudriusButkevicius** [comment]
This doesn't really help with the problem of bootstrapping. I installed syncthing on two devices, I need them to connect/find each other. We don't share a folder, hence this doesn't work.

If the problem is not bootstrapping, i.e., exchange addresses you know for other peers, then in theory you could just add the address sharing at the protocol level, not sure why you'd want to use files.

**Comment by Slender1808** [comment]
my suggestion using file would be but as a first local query if it's not in the list then search on Tracker

**Comment by Slender1808** [comment]
@calmh Marked it as a duplicate but I believe it's different things

**Comment by anacrolix** [comment]
> You might try [anacrolix/dht](https://github.com/anacrolix/dht) and [BEP 44 "Storing arbitrary data in the DHT"](https://www.bittorrent.org/beps/bep_0044.html). @mengzhuo kindly mentioned [anacrolix/torrent](https://github.com/anacrolix/torrent) above, the DHT component has since been broken out into a standalone module.

BEP 44 is now implemented and in use by other projects.

**Comment by calmh** [comment]
@anacrolix Can I somehow use that to announce a set of address strings for a given device ID, and correspondingly look up that set of address strings from somewhere else? I looked through the API docs but it didn't jump out at me.

**Comment by anacrolix** [comment]
Yes, you would hash your device ID, and announce addresses to that hash, it's the typical use case for DHT. It's essentially a `map[[20]byte][]struct{net.IP; uint16}`.

**Comment by AudriusButkevicius** [comment]
How do you prevent spoofing/denial of service?

**Comment by Slender1808** [comment]
> Como você evita spoofing / negação de serviço?

it could have a hash system similar to the one in the emails

**Comment by bt90** [comment]
> How do you prevent spoofing/denial of service?

BEP 44 allows signing:

> It supports both storing immutable items, where the key is the SHA-1 hash of the data itself, and **mutable items, where the key is the public key of the key pair used to sign the data.**

But i'm not sure if that is useful for our usecase as we would end up with another keypair for each client if we can't reuse our existing one for this.

**Comment by calmh** [comment]
> Yes, you would hash your device ID, and announce addresses to that hash, it's the typical use case for DHT. It's essentially a `map[[20]byte][]struct{net.IP; uint16}`.

Keying and security aside, a `[]struct{net.IP; uint16}` doesn't really convey all the info we need (is it tcp? quic? relay?). We could try each for each address, or for IPv4 addresses we could extend them into a fake IPv6 address and use the extra bits to convey protocol information, or we could announce each address as two addresses (one with metadata and one with the actual IP & port in it)...

**Comment by AudriusButkevicius** [comment]
I think that this is implementation specific and you can choose what to store and how. I guess the idea is that every participant should use the same rules for validating.

**Comment by calmh** [comment]
Aight. Cool. Someone should give it a shot. :)

**Comment by bt90** [comment]
We'd still need to figure out how to deal with ed25519 keypairs/signing as this seems mandatory for mutable items.

**Comment by calmh** [comment]
Right; especially since the key we'd be storing things under is the SHA1 hash of that public key. I don't see an obvious way to tie that to a device ID.

**Comment by bt90** [comment]
We might actually end up with a dedicated discovery keypair as BEP 44 mandates only ed25519. But that would defeat the purpose of the device ID IMHO.

**Comment by calmh** [comment]
Not to mention that SHA-1 for certificate use has been deprecated for the last five years or so, it doesn't seem like a forward looking thing to start implementing today. Once could argue that it's just for discovery and the connection is anyway secured properly later, but then you'd still need to communicate this discovery-only ID somehow. (I don't think SHA-1 is broken *enough* that we can easily generate keypairs with a desirable SHA-1 up front. :)

**Comment by bt90** [comment]
This DHT implementation is used by the [Jami](https://jami.net/)(formerly know as Ring)

https://github.com/savoirfairelinux/opendht

It also allows to store arbitrary data but with optional signing:

https://github.com/savoirfairelinux/opendht/wiki#major-features

**Comment by lynzrand** [comment]
Just wondering, does `libp2p` (<https://libp2p.io/>) fit the needs here? It does DHT, record storing, etc., comes from the IPFS project, and has a pretty complete Go implementation. The downside is that its documentation is not very complete, and is still missing some parts (like the NAT traversal feature is not yet stable).

**Comment by calmh** [comment]
@lynzrand maybe, someone should investigate.

@Slender1808 no, because the device ID is already a hash of something else.

**Comment by Slender1808** [comment]
Sorry don't know I couldn't hash hash.

I found this lib maybe useful. 
https://github.com/webtorrent/bittorrent-dht

**Comment by calmh** [comment]
I don't see how it could be, since it's written in Javascript and we've just come to the conclusion that the BitTorrent DHT BEP 44 protocol thing doesn't really work for our purposes.

**Comment by Slender1808** [comment]
about that my suggestion of having table id and local ip on disk synchronizes?

**Comment by calmh** [comment]
It's not a method that can help establish a connection when you're not already connected, so it's not a discovery method. Maybe I set it as a duplicate of the wrong issue. That devices who are already connected could gossip about other devices' addresses is true, and I think we have that considered somewhere, maybe even some code for it. I don't think it needs to be represented as files in folders.

**Comment by Slender1808** [comment]
I had imagined using folders and files to enjoy what is already done

**Comment by anacrolix** [comment]
If the device IDs are secret, and you trust everyone that knows it, use a hash of that as the private key in BEP 44. You're also free to use shared secrets in the salt, as well as in encrypting the data payloads.

**Comment by bt90** [comment]
That could indeed work. Other peers could also verify it by deriving the public key from the private key(device id).

**Comment by calmh** [comment]
Device IDs are explicitly designed to be non-secret. Being able to look up a device ID should not mean the ability to alter their discovery info.

**Comment by bt90** [comment]
Than i'm not sure if this can be solved without introducing at least a shared secret. If we use device ID + shared secret as our private key, an attacker would not be able to guess the DHT key. We would also be able to guard the integrity of the signed data from a malicious node.

Even if we move away from BEP 44 and evaluate other DHT implementations we will eventually face the same problem.

**Comment by calmh** [comment]
I mean, I can imagine a DHT solution where the key is the hash of an arbitrary kind of TLS certificate, and where updates are protected by being signed by that TLS certificate. That's just not what BEP 44 happens to be.

**Comment by bt90** [comment]
Your proposal looks like it is based on the existing certificate+key the device id is based on. Wouldn't this make bootstrapping impossible because the other node only has the device ID which is just a hash of the certificate?

I'm not sure if we even need this level of security *by default*. Knowledge of the device id is enough to publish entries in the global discovery server. Having to know the device id makes not that easy for an attacker and even if succeds to publish his own IPs, syncthing would simply refuse to connect. The worst case is that the attacker could disrupt global discoverabilty for the given node.

My suggestion would be to roll with BEP 44 based on the device id and an **optional** secret to guard it. The private key would then be derived from applying HKDF/PBKDF2 to the device id and the optional secret. If the secret is not set, other nodes are still able to discover the node purely based on its device id but the entry can be overwritten by an attacker who knows the device id and how to derive the keypair from it.

Specifing the DHT secret would be per node, similarly to how we setup untrusted nodes. Discoverying and editing DHT entries only works with the secret. Even if an attacker gets his hands on the device id, he can't lookup or edit any entries associated with it.

But it's not a zero trust model. Other peers are still able to modify entries if they want. And this also needs to be setup by the user on a per node basis which isn't exactly the most comfortable solution.

**Comment by AudriusButkevicius** [comment]
That is strictly worse than what we have now. 

Currently each announcement to discovery server is effectively signed, as the client is required to present its certificate (via tls "client auth required") to the discovery upon announcement, which the discovery server then uses to derive the device id the announcement is for.

So you cannot spoof/dos someone elses addressrs unless you genuinely have the private key.

What we really need is to change what device id means. 

Back when syncthing started ecdsa keys were just coming into fashion, and we used large RSA keys because there was nothing better. It was unfeasible to encode the public key into a device id.

Nowadays, ecdsa keys are small and public part could fit into the device id, allowing us to do things properly, but would obviously require syncthing v2 or some pseudo automated long tail process (have two keys, choose with one to use based on whether we've told the other device that we are changing the key).

Alternative would be to implement a key server (potentially even part of the dht), i.e, you "post" your device id->pub key to the dht, proving that you own that id by signing current time or something, then you can post your device id->addresses signed. Any dht node accepting the addresses announcement reads the pub key and verifies the signature.

You could just send the pub key with the announcements.

**Comment by bt90** [comment]
Migrating from device ids to using device (public) keys is probably the best solution.

The key publishing approach would require a custom DHT implementation/fork as the validation requires Syncthing specific logic in order to check if the device id matches to a public key. Most forms of existing DHT entry signing/verification requires that the value was signed with the key the entry is stored with.

If we use some form of standalone keyserver, we would loose the benefits of DHT. With DHT i can connect to **any** node for bootstrapping and things will work.

What if we really use a distinct keypair for DHT? If nodes exchange their public DHT keys, it would be sufficient if they establish a connection once and persist the public key. The global discovery server would then only be used to bootstrap the very first connection. After that the nodes can simply switch to DHT.

We wojld also avoid tying our device keypair algorithm to the one used by the DHT we pick.

**Comment by calmh** [comment]
There's also the question of what problem is being solved by the new mechanism. I'm aware of two problems with the current centralized infrastructure:

- availability, in that someone could attack it to take it down, or we could choose to cease operating it, leaving users stranded
- confidentiality, in that the entity who runs the centralized infrastructure has access to a top level view of the network (we can count the number of active devices and enumerate them)

A new mechanism that still requires us to run a root node of some kind or, worse, requires devices to register with that root node to enable initial discovery, doesn't seem like much of an improvement.

**Comment by calmh** [comment]
Actually a separate key for DHT, and something as part of the handshake to prove that a device owns that key, would probably work. It would mean a new handshake mechanism and a separate device ID system, none of which is super trivial to add (we're no longer talking about just a discovery mechanism), but it would be doable I think.

**Comment by bt90** [comment]
It would certainly improve availabilty for all devices which have established a connection at some point. We should also keep in mind that Syncthing should still be able to operate in more restrictive networks which might block anything else than 443/tcp. This works fine with global discovery + relays. As DHT is mostly UDP based and often linked to Bittorrent, i'd expect it to be blocked on guest wifi or public hotspots.

We could also allow users to fill in the DHT public key which would allow them to complete the initial setup without global discovery.

**Comment by calmh** [comment]
To be clear, I think any new discovery, DHT based or otherwise, needs to *replace* the current global discovery mechanism (after a multi year migration period etc). Going to all the effort to implement something new only to have to keep running the old infrastructure and have everyone register with it just in case is not great.

**Comment by bt90** [comment]
For DHT we would still need some kind of (selfhostable) HTTP endpoint which simply translates GET/POST to equivalent DHT operations. But this should be opt-in as most users will be able to directly interact with the DHT.

**Comment by bt90** [comment]
This is sadly not limited to DHT. Any discovery mechanism which doesn't provide an (optional) HTTPS endpoint via 443/tcp will render global discovery unuseable for people on restricted networks. This might also include certain carriers or ISPs. We should keep [protocol ossification](https://en.m.wikipedia.org/wiki/Protocol_ossification) and very tight firewall rules in mind with the new design.

IMHO decentraliced discovery should become the new default and the HTTPS based approach should be treated as a fallback users can opt-in to if their network situation requires it. With a selfhosted endpoint they don't even need to rely on the Syncthing project not to murder their hamster.

**Comment by EternityForest** [comment]
@bt90 OpenDHT already specifies an HTTP proxy protocol that's extremely easy to use and works exactly like you sketched.


One thing that I think is worth mentioning that I do when using DHT is rolling codes.   Rather than just hashing the ID, round the current time to a a few hours or ten minutes, and hash along with that to get the key.

That way you never reveal any fixed identifier to random DHT participants, which could then be used for DoS purposes.

**Comment by bt90** [comment]
OpenDHT sounds nice but there seems to be no existing go library for it and we would still need to switch from device ids to full blown public keys to make use of signing.

Migrating Syncthing to use ED25519 would allow us to use it for BEP 44 based DHT discovery.

**Comment by bt90** [comment]
I start to doubt that it is good idea to mix DHT and connection crypto. We would need to create new keypairs for existing installations if we would switch to ED25519. The users would either have to relink their nodes or we end up with a very complex migration path. The new device id would also either need to be the full public key or a hash which fits the DHT implementation we pick in order to guarantee discoverability. As this is tightly coupled to the DHT implementation, we can't simply switch the algorithm in the future like we did in the past when moving from RSA to ECDSA.

Maybe it's easier to add a dedicated discovery ID and figure out how to handle this conviently in the UI. This would also greatly simplify the migration path. Every node would generate a discovery keypair and could exchange its discovery id with already connected nodes.

With BEP 44 as an example, the lookup of a node by its discovery ID would be a two step process:

1. Lookup the public key stored as the value of an immutable item in the DHT with the discovery id(SHA1 hash) as its key
2. Retrieve the addresses stored as a mutable item signed with the public key

We could also additionally store the device ID as part of the mutable item. A client would than only need the discovery id to retrieve all relevant information to establish a secure connection.

**Comment by lynzrand** [comment]
> With BEP 44 as an example, the lookup of a node by its discovery ID would be a two step process: ...

Speaking of this, maybe we should try [libp2p's DHTs][libp2p-dht], which has 256-bit keys and a dedicated peer address storage. In this way we can just reuse our device ids if we want.

Another idea is that, instead of relying on signatures on the addresses in the network, we just connect to those endpoints, let them send their certificates to us before we send ours to them (the TLS protocol seems to do this already), and just authenticate those certificates using the device id. This differs from the current Block Exchange Protocol since the connection can be aborted _before_ the TLS handshake completes, but it shouldn't be a big deal (`crypto/tls/Config` has a callback field `GetClientCertificate` so we can implement this authentication; and to the otherside of the handshake, it should just look like a regular TLS abort).

[libp2p-dht]: https://pkg.go.dev/github.com/libp2p/go-libp2p-kad-dht

**Comment by bt90** [comment]
libp2ps DHT doesn't seem to be general purpose: https://github.com/libp2p/go-libp2p-kad-dht/issues/597

The problem with your approach was already pointed out by @AudriusButkevicius. If entries are not guarded by public key cryto, anyone can overwrite entries with garbage and disrupt global discovery.

**Comment by lynzrand** [comment]
> libp2ps DHT doesn't seem to be general purpose

Sorry about that. I was primarily reading the Rust side of libp2p's implementation, and thought that the Go side would be the same.

> If entries are not guarded by public key cryto, anyone can overwrite entries with garbage and disrupt global discovery.

(Correct me if I'm wrong) 

I'm thinking of saving these entries as peers that nodes connect to, instead of data entries. As far as I know, peer information in DHT is already validated when nodes connect to them. So if a malicious node writes garbage peer information into the network, they simply fail authentication when other nodes try to connect to them (certificate mismatch, etc.), so they won't propagate through the network.

**Comment by bt90** [comment]
Peer info only allows host+port AFAIK. So we would need to guess the protocol(tcp/quic and couldn't store relay addresses. Validation by connecting to the node also doesn't verify any certificates. It's just a simple measure to prevent the DHT getting spammed.

**Comment by bt90** [comment]
I think the blocker is mostly the bootstrapping of the initial connection. The current approach uses the device ID to do this. If you know the ID of a device, you can look it up via the global discovery server.

So far we haven't found a way to mimic this in a DHT, as we also want the published addresses to be authenticated. An attacker should not be able to spoof or manipulate the DHT entries just by knowing the device ID.

DHTs support this kind of authenticated entries but use public key cryptography to implement it. The problem is that the device ID is not a public key but only a fingerprint of the TLS certificate.

So either we switch the protocol from certificates to public keys or come up with our own DHT with a custom entry authentication scheme.

**Comment by SoniEx2** [comment]
you can use a fingerprint-based DHT to store the certificate (aka public key) and use a key-based DHT to store the important bits.

**Comment by bt90** [comment]
The problem is that the other peer doesn't have the certificate. It only has the fingerprint of the certificate.

**Comment by SoniEx2** [comment]
the solution is to store the full certificate in the DHT.

you lookup the fingerprint, you get the certificate.
you lookup the public key, you get the important bits.

**Comment by bt90** [comment]
> you lookup the fingerprint, you get the certificate.

The key of a DHT entry is either the hash of it's content or the public key for an authenticated entry. The fingerprint is not a public key and way too special to align with any algorithm used for content hashing:

https://docs.syncthing.net/dev/device-ids.html#id1

**Comment by SoniEx2** [comment]
It sounds like the fingerprint is the SHA-256 hash of the certificate.

What makes a SHA-256-based immutable DHT inappropriate for that?

**Comment by bt90** [comment]
The device ID format should be fully reversible, so the peer would be able to discover the full certificate *if* the DHT uses SHA256. The remaining issue is the discovery of the authenticated entry. While we do have a public/private key pair, it may not match the algorithm used by the DHT, e.g. BEP44 defines ed25519 public keys for mutable items.

**Comment by SoniEx2** [comment]
what stops you from uh, running multiple DHTs for all the supported key types?

**Comment by sjehuda** [comment]
DHT is one of the technologies that bad governments and bad corporations hate the most.

DHT is an important technology. I operate several machines world wide to share over 5 million files via DHT over the BitTorrent system and I use **not** even a single BitTorrent Tracker to discover peers.

@AudriusButkevicius
> You still need infrastructure to bootstrap, and how do we perform authentication against other peers being non-malicious, prevent malicious peers from overwriting keys, or even bootstrap server from being malicious?

@SoniEx2
> what stops you from uh, running multiple DHTs for all the supported key types?

Gentlemen, these are good notes you have made.

As I recall, [Jami](https://jami.net/) (previously GNU Ring, and SFLphone), from @savoirfairelinux, which has started as a SIP/IAX VoIP client has later added support to a completely decentralized system which uses DHT to discover peers of its decentralized VoIP system.

See [OpenDHT](http://opendht.net/) and 
[this article](https://github.com/syncthing/syncthing/files/15271136/censored_article_about_jami_ring_and_dht.tar.gz) (Tar archive) which has been censored either by an intelligence agency or by a corporation.

Also, to increase identity verification, make DHT possible by using PGP keys, similarly to the [cjdns](https://github.com/cjdelisle/cjdns) system from @cjdelisle.

**Comment by sjehuda** [comment]
You might also want to add support to cjdns and I2P systems.

You can refer to Seth who gracefully explains the I2P system.
- https://fosstodon.org/@RTP
- https://tube.tchncs.de/c/privacy__tech_tips/videos?s=1
- http://righttoprivacy.i2p/

**Comment by sjehuda** [comment]
#### Another important comment about the importance DHT.

The so called entertainment industry (i.e. inteligence controlled cultural subversion industry) has ceased to discuss about P2P for over a decade and instead of publicly fighting BitTorrent.=, and instead the industry has delivered the so called "streaming services".

The reason for this is because DHT has made BitTorrent unstoppable and the best resort to fight against BitTorrent is to stop talking about it and hope people would forget about it.

#### Centralized resources are subjected to compromisation

The Syncthing Relay Nodes, and also BitTorrent Trackers, are centralized in nature and have an identity which can be tracked and potentially hijacked by a bad entity.

**Comment by sjehuda** [comment]
I have added more notes to: ### Centralized resources are more often subjected to compromisation than decentralized resources

**Comment by cjdelisle** [comment]
> DHT is one of the technologies that bad governments and bad corporations hate the most.

It's not quite like that, all DHTs are inherently susceptible to poison and sybil attacks. For example if I don't like entry XXXX I can generate a ton of fake entries and keep the ones with keys which are numerically closest to XXXX, then I insert them into the DHT to clog up the nodes which hold them. Likewise, I can generate fake nodes, using the same collision attack to get node IDs that are numerically closest to XXXX such that most nodes will drop XXXX because it appears that it is already handled by my (fake) nodes.

> The reason for this is because DHT has made BitTorrent unstoppable

Yes and no. The Bittorrent DHT does make Bittorrent a lot more robust against attacks, but not for the reasons you might think. The DHT *is* attackable, as stated, but the nodes in the DHT are not all what they appear to be. The major DHT nodes are not some uTorrent install running on someone's laptop, they're on servers operated by the Good Ol' Boys, so if you start trying to attack the DHT, you may find your IP addresses suddenly getting blocked and then you're unreachable to most of the DHT network. The DHT bootstraps off of DNS seed servers and if the seed operators only point people to the "known good guy" DHT nodes, it's easy for an attacker to just end up isolated from the whole network.

When this is combined with the proliferation of Trackers, it creates a platform which is very frustrating to attack. You can go after the DHT and it will degrade somewhat but you don't get to pull out a clear win, and in any case the trackers will carry the network so basically no one will notice that anything happened. You can also go after the trackers, but there are a lot of them, in a lot of countries, and even if you manage to take down a large subset of trackers, the network will be carried by the DHT, so attacks on Bittorrent are just miserably frustrating to carry out.

> You might also want to add support to cjdns and I2P systems.

For cjdns, all you need is IPv6 support and to *not* forbid communication with addresses in the `fc00::/8` block.

**Comment by calmh** [comment]
There's a whole lot of handwaving in this thread. We have a fairly simple set of requirements:

1. we need to map a 256 bit key (the device ID) to a small piece of data (the discovery information)
2. that information needs to change regularly
3. we want to run a check on that information (e.g., a signature) before accepting it as legitimate

All the "get an immutable piece of data by hash" or "get an ip:port for a node with a random 160 bit ID" solutions don't move anything forward.

**Comment by EternityForest** [comment]
The public key could be included in the DHT payload rather than using the
DHTs native signing, if it's something small like ECC. The DHT key could
just be the hash of the node ID, or time based rolling code hash for more
privacy.

Or the node ID itself could be treated as semi-private, and used to encrypt
and authenticate the payload.

The only consequence to accepting bad data from the DHT would be a
connected attempt to an attacker node, which would fail but reveal your IP
to them(Same as you can already get if you know the node ID), and cause you
to be DoSed, just like you would be with no DHT at all if the primary
servers went down, right? Is my understanding of the threat model accurate?

On home connections you can reuse a cached IP for weeks, so a DHT can still
be far better than nothing even if it's not perfectly reliable.

**Comment by jadsongmatos** [comment]
One of the ways to reduce server waste would be to leverage existing Torrent Tracker servers, similar to the P2PT project, which uses WebTorrent Trackers as signaling servers for P2P connections.

P2PT is a library that connects P2P very differently. P2PT works by using WebTorrent Trackers as signaling servers, converting an application's unique identifier into a valid Info Hash, which is sent to the trackers to obtain a list of web peers using the application.

By taking a similar approach, Syncthing could use WebTorrent Trackers for P2P discovery, reducing reliance on dedicated servers and improving scalability. This method uses existing infrastructure to reduce costs and make communication more efficient.

**References:**
- P2PT API documents: https://github.com/subins2000/p2pt/blob/master/api-docs.md
- WebTorrent: https://webtorrent.io/
- Magnetic URI scheme: https://en.wikipedia.org/wiki/Magnet_URI_scheme#URN,_containing_hash_(xt)

**Comment by aydencook03** [comment]
A possible alternative to DHT that still keeps discovery decentralized would be to use Nostr relays. There are many p2p projects that utilize nostr ephemeral events for peer discovery. [Hashtree](https://hashtree.cc/#/dev:~:text=via%20Nostr%20ephemeral%20events) is an example. This would be simpler and more lightweight than a DHT.

**Comment by sjehuda** [comment]
@aydencook03 This might be a good idea. Do you have code references?

Nevertheless, DHT is probably preferable, as it is utilized with BitTorrent, eDonkey2000, Gnutella, and other networks; and, most implementations are lean.

Of course, implementing a mechanism to query consequent IP addresses (so called "brute force") be a good addition to software which support DHT.

Perhaps implementing all three (DHT, Nostr, Query) be a good idea.

**Comment by sjehuda** [comment]
@aydencook03 This might be a good idea. Do you have code references?

**Comment by aydencook03** [comment]
@sjehuda I don't. But if people think it's a good idea then I could find some or put something together myself. I've made a [post](https://forum.syncthing.net/t/nostr-for-peer-discovery-instead-of-dht/26616) in the forums so that this idea can be discussed.

**Comment by sjehuda** [comment]
While I prefer and biased towards DHT, as I have experience with DHT, your idea might be even better.

The fact that maybe it was not tried yet, does not mean that it is not better.

I am not an authority; yet,, I am infavour of considering your idea.
